### PR TITLE
refactor(@angular-devkit/build-angular): clean find tests logic

### PR DESCRIPTION
**Pull request by alan-agius4**
With this change we clean up and reduce the complexity of the `findTests` logic that handles the `include` option in the karma builder.

### CODE REVIEW

File: packages/angular_devkit/build_angular/src/builders/karma/find-tests.ts [hunk 1]
```diff
@@ -6,55 +6,85 @@
...
 }
 
-function findMatchingTests(pattern: string, cwd: string, relativePathToMain: string): string[] {
+const normalizePath = (path: string): string => path.replace(/\\/g, '/');
+
+async function findMatchingTests(
+  pattern: string,
+  workspaceRoot: string,
+  projectSourceRoot: string,
+): Promise<string[] | string> {
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/builders/karma/find-tests.ts:37]: Can we have this always return `string[]` and just collapse a single result to `[ result ]`?

File: packages/angular_devkit/build_angular/src/builders/karma/find-tests.ts [hunk 2]
```diff
@@ -6,55 +6,85 @@
...
+): Promise<string[]> {
+  const matchingTestsPromises = patterns.map((pattern) =>
+    findMatchingTests(pattern, workspaceRoot, projectSourceRoot),
+  );
+  const files = await Promise.all(matchingTestsPromises);
 
-    return files;
-  }, [] as string[]);
+  // Unique file names
+  return [...new Set(files.flatMap((f) => f))];
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/builders/karma/find-tests.ts:28]: Could we use `flies.flat()` here? An empty `flatMap()` is a bit unintuitive in its purpose.

File: packages/angular_devkit/build_angular/src/builders/karma/index.ts [hunk 3]
```diff
@@ -112,9 +112,22 @@ export function execute(
...
+        if (!projectName) {
+          throw new Error('The builder requires a target.');
+        }
+
+        const projectMetadata = await context.getProjectMetadata(projectName);
+        const projectSourceRoot = getSystemPath(
+          join(
+            normalize(context.workspaceRoot),
+            (projectMetadata.root as string | undefined) || '',
+            (projectMetadata.sourceRoot as string | undefined) || '',
```
Comment (dgp1130) [inline packages/angular_devkit/build_angular/src/builders/karma/index.ts:126]: Would `??` work here and better communicate the intent?

**Review by dgp1130 — APPROVED** [review]
Just a few coding nitpicks.

File: packages/angular_devkit/build_angular/src/builders/karma/find-tests.ts (hunk 2 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/builders/karma/find-tests.ts:28]: Good call!

**Pull request merged**